ballerina/auth_desugar.bal

```diff
@@ -0,0 +1,162 @@
+// UnauthenticatedError/ PermissionDeniedError sent by the `grpc:Caller` which is taken with an interop call. The
+// execution flow will be broken by panic with a distinct error.
+# Uses for declarative auth design, where the authentication/authorization decision is taken
+# by reading the auth annotations provided in service/resource and the `Authorization` header of request.
+# 
+# + serviceRef - The service reference where the resource locates
+# + methodName - The name of the subjected resource
+# + resourcePath - The relative path
+public isolated function authenticateResource(Service serviceRef) {
+    ListenerAuthConfig[]? authConfig = getListenerAuthConfig(serviceRef);
+    if authConfig is () {
+        return;
+    }
+    string|Error authHeader = externGetAuthorizationHeader();
+    if authHeader is string {
+        map<string> headers = {
+            "authorization": authHeader
+        };
+        UnauthenticatedError|PermissionDeniedError? result = tryAuthenticate(<ListenerAuthConfig[]>authConfig, headers);
+        if result is UnauthenticatedError || result is PermissionDeniedError {
+            sendError(result);
+        }
+    } else {
+        sendError(authHeader);
+    }
+}
+
+isolated function tryAuthenticate(ListenerAuthConfig[] authConfig, map<string|string[]> headers) returns UnauthenticatedError|PermissionDeniedError? {
+    PermissionDeniedError|boolean? result = ();
+    foreach ListenerAuthConfig config in authConfig {
+        if config is FileUserStoreConfigWithScopes {
+            result = authenticateWithFileUserStore(config, headers);
+        } else if config is LdapUserStoreConfigWithScopes {
+            result = authenticateWithLdapUserStore(config, headers);
+        } else if config is JwtValidatorConfigWithScopes {
+            result = authenticateWithJwtValidator(config, headers);
+        } else {
+            result = authenticateWithOAuth2IntrospectionConfig(config, headers);
+        }
+        if result is PermissionDeniedError {
+            return result;
+        } else if result is boolean && result {
+            return;
+        }
+    }
+    return error UnauthenticatedError("Failed to authenticate client");
+}
+
+isolated function authenticateWithFileUserStore(FileUserStoreConfigWithScopes config, map<string|string[]> headers)
+ returns PermissionDeniedError|boolean? {
+    ListenerFileUserStoreBasicAuthHandler handler = new(config.fileUserStoreConfig);
+    auth:UserDetails|UnauthenticatedError authn = handler.authenticate(headers);
+    string|string[]? scopes = config?.scopes;
+    if authn is auth:UserDetails {
+        if scopes is string|string[] {
+            PermissionDeniedError? authz = handler.authorize(authn, scopes);
+            if authz is PermissionDeniedError {
+                return authz;
+            }
+            return true;
+        }
+        return true;
+    }
```

Comment:
@dilanSachi Shall we simplify the logic as follows, so that we can remove returning `boolean` just for the sake of handling the logic of parent method? With this approach the return values of this method would be `UnauthenticatedError| PermissionDeniedError?`. Parent method can simply call this function and return whatever the result. WDYT?
```suggestion
    if authn is auth:UserDetails {
        if scopes is string|string[] {
            PermissionDeniedError? authz = handler.authorize(authn, scopes);
            return authz;
        }
    }
    return error UnauthenticatedError(UNAUTHENTICATED_ERROR_MSG);
```

Reply:
With this, if `authn` is not `auth:UserDetails` or `scopes` is not `string|string[]`, the loop will continue without authorizing. I had done what is mentioned here just as how `http` has [done](https://github.com/ballerina-platform/module-ballerina-http/blob/14bb854389a13d893ef667ae427c51b212cf8cdd/ballerina/auth_desugar.bal#L56). But after dividing it to funtions, I had to use a boolean.

Reply:
Okay. I got your point. I think that was the reason that we didn't split into methods. Anyway, shall we simplify as follows? With this approach the return values of this method would be `PermissionDeniedError|boolean`. Parent method can simply call this function and return the `PermissionDeniedError` by using `check` expression and the remaining `boolean` value will be assigned to the `result` variable. At the stage of analyzing the `boolean`, we can check for `true` and return, if `false`, we can continue the loop. WDYT?
```suggestion
    if authn is auth:UserDetails {
        if scopes is string|string[] {
            PermissionDeniedError? authz = handler.authorize(authn, scopes);
            if authz is PermissionDeniedError {
                return authz;
            }
            return true;
        }
        return true;
    }
    return false;
```

Reply:
Agreed. Will change accordingly

Reply:
Changed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/343/commits/2b5be7c79b1b2e947c5351acc96d25881ef65b10

Reply:
Changed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/343/commits/17bcfaa6a8dcf712b24495b94565a75c19219756